parachain/pallets/dot-app/src/lib.rs

```diff
@@ -0,0 +1,116 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+use frame_system::{self as system, ensure_signed};
+use frame_support::{
+	decl_error, decl_event, decl_module, decl_storage,
+	dispatch::{DispatchError, DispatchResult},
+	traits::{
+		Get,
+		EnsureOrigin,
+		Currency,
+		ExistenceRequirement::{KeepAlive, AllowDeath},
+	}
+};
+use sp_runtime::traits::StaticLookup;
+use sp_std::prelude::*;
+use sp_core::H160;
+use sp_runtime::{
+	ModuleId,
+	traits::AccountIdConversion,
+	SaturatedConversion,
+};
+
+use artemis_core::{ChannelId, SubmitOutbound};
+
+mod payload;
+use payload::OutboundPayload;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+type BalanceOf<T> = <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+pub trait Config: system::Config {
+	type Event: From<Event<Self>> + Into<<Self as system::Config>::Event>;
+
+	type Currency: Currency<Self::AccountId>;
+
+	type SubmitOutbound: SubmitOutbound;
+
+	type CallOrigin: EnsureOrigin<Self::Origin, Success=H160>;
+
+	type ModuleId: Get<ModuleId>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Config> as EthModule {
+		/// Address of the peer application on the Ethereum side.
+		Address get(fn address) config(): H160;
+	}
+}
+
+decl_event!(
+    /// Events for the ETH module.
+	pub enum Event<T>
+	where
+		AccountId = <T as system::Config>::AccountId,
+		Balance = BalanceOf<T>
+	{
+		Locked(AccountId, H160, Balance),
+		Unlocked(H160, AccountId, Balance),
+	}
+);
+
+decl_error! {
+	pub enum Error for Module<T: Config> {
+
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: T::Origin {
+
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		#[weight = 0]
+		pub fn lock(origin, channel_id: ChannelId, recipient: H160, amount: BalanceOf<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			T::Currency::transfer(&who, &Self::account_id(), amount, AllowDeath)?;
+
+			let message = OutboundPayload {
+				sender: who.clone(),
+				recipient: recipient.clone(),
+				amount: amount.saturated_into::<u128>(),
+			};
+
+			T::SubmitOutbound::submit(channel_id, Address::get(), &message.encode().unwrap())?;
```

Comment:
Instead of panicing here, can we return the error?

Reply:
Also, please encode and return errors first before doing the transfer and the submit. Substrate does not roll back the application state, so an error in message encoding would lead to a transfer without a submit.

Reply:
good catch! Not sure how I overlooked that. Will add a fix.

Reply:
Fixed now. I've wrapped both the `unlock` and `lock` calls with the [transactional](https://substrate.dev/rustdocs/v3.0.0/frame_support/attr.transactional.html) macro.

I've also added a test in [6d1513](https://github.com/Snowfork/polkadot-ethereum/commit/6d15131e28bf266e32a13701c7523554e6b85dec) to verify that if `T::SubmitOutbound::submit()` fails, then any previous storage changes (such as the asset transfer) would be rolled back.